[DebugInfo] Fixed default emission of ".debug_names" section

SouraVX opened the pull request:
Metdata field "nameTableKind" is responsible for generation/emission of
".debug_names" section.
Set it appropriately to avoid generation/emission of ".debug_names" section by default.

SouraVX commented:
Driver side changes(https://github.com/flang-compiler/flang-driver/pull/95) has to be merged first get appropriate functionality.

gklimowicz reviewed (approved):
This passes our tests on OpenPOWER for LLVM 9 and LLVM 10.

SouraVX commented:
@gklimowicz could you please also try Driver side changes flang-compiler/flang-driver#95

gklimowicz commented:
> @gklimowicz could you please also try Driver side changes [flang-compiler/flang-driver#95](https://github.com/flang-compiler/flang-driver/pull/95)

@SouraVX Will do. Sorry I missed the note about[flang-compiler/flang-driver#95](https://github.com/flang-compiler/flang-driver/pull/95)

gklimowicz commented:
This fails the flang lit test `debug_info/debug_module_import.f90` on LLVM 9 and LLVM 10 on OpenPOWER. I didn't check on other platforms.

Below is the output from `make check-clang`:
```
FAIL: Flang :: debug_info/debug_module_import.f90 (16 of 1350)
******************** TEST 'Flang :: debug_info/debug_module_import.f90' FAILED ********************
Script:
--
: 'RUN: at line 17';    /home/gak/workspaces/flang-pr-process/build.pr934.ppc64le.90/install/bin/flang  -g -S -emit-llvm /home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90 -o - | /home/gak/workspaces/flang-pr-process/build.pr934.ppc64le.90/flang/bin/FileCheck /home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90
--
Exit Code: 1

Command Output (stderr):
--
/home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90:38:12: error: CHECK: expected string not found in input
! //CHECK: ![[DBG_DIC:[0-9]+]] = distinct !DICompileUnit({{.*}}, imports: ![[DBG_IMPORTS:[0-9]+]])
           ^
<stdin>:117:40: note: scanning from here
!2 = !DIModule(scope: !3, name: "first")
                                       ^
<stdin>:118:1: note: possible intended match here
!3 = distinct !DICompileUnit(language: DW_LANG_Fortran90, file: !4, producer: " F90 Flang - 1.5 2017-05-01", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, enums: !5, retainedTypes: !5, globals: !6, imports: !11, nameTableKind: None)
^

--
```

SouraVX commented:
> This fails the flang lit test `debug_info/debug_module_import.f90` on LLVM 9 and LLVM 10 on OpenPOWER. I didn't check on other platforms.
> 
> It did not fail our NVIDIA internal tests that I ran.
> 
> Below is the output from `make check-clang`:
> 
> ```
> FAIL: Flang :: debug_info/debug_module_import.f90 (16 of 1350)
> ******************** TEST 'Flang :: debug_info/debug_module_import.f90' FAILED ********************
> Script:
> --
> : 'RUN: at line 17';    /home/gak/workspaces/flang-pr-process/build.pr934.ppc64le.90/install/bin/flang  -g -S -emit-llvm /home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90 -o - | /home/gak/workspaces/flang-pr-process/build.pr934.ppc64le.90/flang/bin/FileCheck /home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90
> --
> Exit Code: 1
> 
> Command Output (stderr):
> --
> /home/gak/workspaces/flang-pr-process/src.pr934.90/flang/test/debug_info/debug_module_import.f90:38:12: error: CHECK: expected string not found in input
> ! //CHECK: ![[DBG_DIC:[0-9]+]] = distinct !DICompileUnit({{.*}}, imports: ![[DBG_IMPORTS:[0-9]+]])
>            ^
> <stdin>:117:40: note: scanning from here
> !2 = !DIModule(scope: !3, name: "first")
>                                        ^
> <stdin>:118:1: note: possible intended match here
> !3 = distinct !DICompileUnit(language: DW_LANG_Fortran90, file: !4, producer: " F90 Flang - 1.5 2017-05-01", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, enums: !5, retainedTypes: !5, globals: !6, imports: !11, nameTableKind: None)
> ^
> 
> --
> ```

This seems not relevant to this patch(at least). By any chance, can it be a random failure ?

kiranchandramohan reviewed (commented):
Not clear what is the relation of this PR with the flang-driver #95 PR. https://github.com/flang-compiler/flang-driver/pull/95

1. Why is the gpubnames flag not used for the test?
2. Why is the magic number 0x40000000 not used?
3. Can we have a test with a table not none?

SouraVX commented:
Thanks for review! Addressed comments by @kiranchandramohan .
- Modified test to check/test both scenarios presence/absence of nameTable based on commandline flang `-gpubnames`. (Note: Driver PR has to be applied first to get this test case passing).

SouraVX commented:
Ping!

michalpasztamobica reviewed (changes requested):
I tried this out with the changes from https://github.com/flang-compiler/classic-flang-llvm-project/pull/25, but I still got the test to fail...
I tried on release_100 and the upcoming release_12x (I have some local tools conflict when trying with release_11x) and each time got this:

```
/home/mipas/flang/flang/test/debug_info/nametable.f90:8:11: error: NOTABLE: expected string not found in input
!NOTABLE: !DICompileUnit({{.*}}, nameTableKind: None
          ^
<stdin>:1:1: note: scanning from here
; ModuleID = '/tmp/nametable-b395a3.ll'
^
<stdin>:28:15: note: possible intended match here
!2 = distinct !DICompileUnit(language: DW_LANG_Fortran90, file: !3, producer: " F90 Flang - 1.5 2017-05-01", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, enums: !4, retainedTypes: !4, globals: !4, imports: !4)
              ^

Input file: <stdin>
Check file: /home/mipas/flang/flang/test/debug_info/nametable.f90
```

Would you please fix the missing comma and try to run the tests to see if they also fail on your side or if it is something I am doing wrong?
Thanks!

SouraVX commented on tools/flang2/flang2exe/ll_write.cpp:
```diff
@@ -871,6 +871,7 @@ enum FieldType {
   DWEncodingField,
   DWEmissionField,
   SignedOrMDField
```
Thank You so much for catching this.

SouraVX commented:
>Would you please fix the missing comma and try to run the tests to see if they also fail on your side or if it is something I am doing wrong?
Thanks!

I crossed checked on my local machine with https://github.com/flang-compiler/classic-flang-llvm-project/pull/25 as parent or pre-requisite patch.
My builds are clean, Test case is passing and desired functionality is getting achieved.

michalpasztamobica commented:
Hi @SouraVX ,
So I tried again and again, but kept getting failures, so in the end I decided to try out your patches on the actual CI.
I built the `classic-flang-llvm-project` changes [on my fork](https://github.com/michalpasztamobica/classic-flang-llvm-project/actions/runs/687276748) by cherry-picking them to my `release_11x` branch.
Then I cherry-picked your `flang` changes and limited the amount of builds running to only `release_11x`.
With this setup, which is exeactly how our CI works I managed to reproduce the [problem](https://github.com/michalpasztamobica/flang/runs/2196122870?check_suite_focus=true). 
Would you have a look at this error report?
It actually does originate from the newly added test:
```
Testing: 
FAIL: Flang :: debug_info/nametable.f90 (21 of 1376)
******************** TEST 'Flang :: debug_info/nametable.f90' FAILED ********************
Script:
--
: 'RUN: at line 1';    /usr/local/bin/flang  /home/runner/work/flang/flang/test/debug_info/nametable.f90 -gdwarf-5 -S -emit-llvm -o - | /home/runner/work/flang/flang/build/bin/FileCheck --check-prefix=NOTABLE /home/runner/work/flang/flang/test/debug_info/nametable.f90
: 'RUN: at line 2';    /usr/local/bin/flang  /home/runner/work/flang/flang/test/debug_info/nametable.f90 -g -S -emit-llvm -o - | /home/runner/work/flang/flang/build/bin/FileCheck --check-prefix=NOTABLE /home/runner/work/flang/flang/test/debug_info/nametable.f90
: 'RUN: at line 4';    /usr/local/bin/flang  /home/runner/work/flang/flang/test/debug_info/nametable.f90 -gdwarf-5 -gpubnames -S -emit-llvm -o - | /home/runner/work/flang/flang/build/bin/FileCheck --check-prefix=TABLE /home/runner/work/flang/flang/test/debug_info/nametable.f90
: 'RUN: at line 5';    /usr/local/bin/flang  /home/runner/work/flang/flang/test/debug_info/nametable.f90 -g -gpubnames -S -emit-llvm -o - | /home/runner/work/flang/flang/build/bin/FileCheck --check-prefix=TABLE /home/runner/work/flang/flang/test/debug_info/nametable.f90
--
Exit Code: 1

Command Output (stderr):
--
/home/runner/work/flang/flang/test/debug_info/nametable.f90:11:13: error: TABLE-NOT: excluded string found in input
!TABLE-NOT: !DICompileUnit({{.*}}, nameTableKind: None
```
